server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1264,25 +1264,45 @@ private ReaderContext findReaderContext(ShardSearchContextId id, TransportReques
     }
 
     final ReaderContext createOrGetReaderContext(ShardSearchRequest request) {
-        ShardSearchContextId readerId = request.readerId();
-        if (request.readerId() != null) {
+        ShardSearchContextId contextId = request.readerId();
+        final long keepAliveInMillis = getKeepAlive(request);
+        if (contextId != null) {
             try {
-                return findReaderContext(readerId, request);
+                return findReaderContext(contextId, request);
             } catch (SearchContextMissingException e) {
-                if (readerId.isRetryable() == false) {
+                if (contextId.isRetryable() == false) {
                     throw e;
                 }
+                // We retry creating a ReaderContext on this node if we can get a searcher with same searcher id as in the original
+                // PIT context. This is currently possible for e.g. ReadOnlyEngine and its child FrozenEngine where a commitId is
+                // calculated from the ids of the underlying segments of an index commit
                 final IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
                 final IndexShard shard = indexService.getShard(request.shardId().id());
                 final Engine.SearcherSupplier searcherSupplier = shard.acquireSearcherSupplier();
-                if (readerId.sameSearcherIdsAs(searcherSupplier.getSearcherId()) == false) {
+                if (contextId.sameSearcherIdsAs(searcherSupplier.getSearcherId()) == false) {
                     searcherSupplier.close();
                     throw e;
                 }
-                return createAndPutReaderContext(request, indexService, shard, searcherSupplier, defaultKeepAlive);
+                ReaderContext readerContext = null;
+                if (PIT_RELOCATION_FEATURE_FLAG.isEnabled()) {
+                    // we are using a PIT here so set singleSession to false to prevent clearing after the search finishes
+                    readerContext = createAndPutReaderContext(
+                        contextId,
```

Comment:
I think we should create a new contextId instead of reusing it?

Reply:
The sessionId of this contextId can be different from the sessionId of the SearchService.